csharp/src/Ice/UriParser.cs

```diff
@@ -0,0 +1,340 @@
+                        facet,
+                        invocationMode,
+                        path,
+                        protocol);
+            }
+            catch (Exception ex)
+            {
+                // Give context to the exception.
+                throw new FormatException($"failed to parse URI `{uriString}'", ex);
+            }
+        }
+
+        /// <summary>Registers the ice and ice+universal schemes.</summary>
+        internal static void RegisterCommon()
+        {
+            RegisterTransport("universal", defaultPort: 0);
+
+            // There is actually no authority at all with the ice scheme, but we emulate it with an empty authority
+            // during parsing by the Uri class and the GenericUriParser.
+            GenericUriParserOptions options =
+                ParserOptions |
+                GenericUriParserOptions.AllowEmptyAuthority |
+                GenericUriParserOptions.NoPort;
+
+            System.UriParser.Register(new GenericUriParser(options), "ice", -1);
+        }
+
+        /// <summary>Registers an ice+transport scheme.</summary>
+        /// <param name="transportName">The name of the transport (cannot be empty).</param>
+        /// <param name="defaultPort">The default port for this transport.</param>
+        internal static void RegisterTransport(string transportName, ushort defaultPort) =>
+            System.UriParser.Register(new GenericUriParser(ParserOptions), $"ice+{transportName}", defaultPort);
+
+        private static Endpoint CreateEndpoint(
+            Communicator communicator,
+            bool oaEndpoint,
+            Dictionary<string, string> options,
+            Protocol protocol,
+            Uri uri,
+            string uriString)
+        {
+            string transportName = uri.Scheme.Substring(4); // i.e. chop-off "ice+"
+
+            // TODO: take advantage of other properties of uri such as HostNameType?
+            string host = uri.Host;
+            if (host[0] == '[' && host[^1] == ']')
+            {
+                // Remove brackets around IPv6 addresses
+                host = host[1..^1];
+            }
+
+            ushort port = 0;
+            checked
+            {
+                port = (ushort)uri.Port;
+            }
+
+            IEndpointFactory? factory = null;
+            Transport transport = default;
+
+            if (transportName == "universal")
+            {
+                if (protocol == Protocol.Ice1)
+                {
+                    throw new FormatException("ice+universal is not compatible with the ice1 protocol");
+                }
+                if (oaEndpoint)
+                {
+                    throw new FormatException("ice+universal cannot specify an object adapter endpoint");
+                }
+
+                // Enumerator names can only be used for "well-known" transports.
+                transport = Enum.Parse<Transport>(options["transport"], ignoreCase: true);
```

Comment:
should we reject numeric values here when parsing the transport?

Reply:
No, we want to be able to specify numeric transports that are not well-known (well-known = there is an enumerator).

Reply:
I think `Enum.Parse` throw if the numeric value is not one of the valid enumerators

Reply:
It does not, and some of the tests check this code.